pioneer/packages/joy-utils/src/transport/base.ts

```diff
@@ -38,4 +43,50 @@ export default abstract class BaseTransport {
   protected get minting () {
     return this.api.query.minting;
   }
+
+  protected queryMethodByName (name: string) {
+    const [module, method] = name.split('.');
+    return this.api.query[module][method];
+  }
+
+  // Fetch all double map entries using only the first key
+  //
+  // TODO: FIXME: This may be a risky implementation, because it relies on a few assumptions about how the data is stored etc.
+  // With the current runtime version we can rely on the fact that all storage keys for double-map values start with the same
+  // 32-bytes prefix assuming a given (fixed) value of the first key (ie. for all values like map[x][y], the storage key starts
+  // with the same prefix as long as x remains the same. Changing y will not affect this prefix)
+  protected async doubleMapEntries<T extends Codec> (
+    methodName: string,
+    firstKey: Codec,
+    valueConverter: (hex: string) => T,
+    getEntriesCount: () => Promise<number>,
+    secondKeyStart = 1
+  ): Promise<{ secondKey: number; value: T}[]> {
+    // Get prefix and storage keys of all entries
+    const firstEntryStorageKey = this.queryMethodByName(methodName).key(firstKey, secondKeyStart);
+    const entryStorageKeyPrefix = firstEntryStorageKey.substr(0, 66); // "0x" + 64 hex characters (32 bytes)
+    const allEntriesStorageKeys = await this.api.rpc.state.getKeys(entryStorageKeyPrefix);
+
+    // Create storageKey-to-secondKey map
+    const maxSecondKey = (await getEntriesCount()) - 1 + secondKeyStart;
+    const storageKeyToSecondKey: { [key: string]: number } = {};
+    for (let secondKey = secondKeyStart; secondKey <= maxSecondKey; ++secondKey) {
+      const storageKey = this.queryMethodByName(methodName).key(firstKey, secondKey);
+      storageKeyToSecondKey[storageKey] = secondKey;
+    }
+
+    // Create the resulting entries array
+    const entries: { secondKey: number; value: T }[] = [];
+    for (const key of allEntriesStorageKeys) {
+      const value: any = await this.api.rpc.state.getStorage(key);
+      if (typeof value === 'object' && value !== null && value.raw) {
```

Comment:
In the past when playing with reading directly from storage I found that basically if you look at the raw bytes, if the first byte is 0 then the key/value pair is not in storage. If it is 1 then it exists and the remaining bytes are the raw data. Is there no "shortcut" property on the return value to do this check in a type safe way? No need to change here just wondering for future.